CORE: MC/UC: reductions refactor

vspetrov opened the pull request:
## What
Moves all the reductions from MC to EC
Refactors EE task interface
Extends EE Task Reduce for multi-buffer support 
Unifies/Simplifies old ucc_mc_reduce, ucc_mc_reduce_multiple, ucc_mc_reduce_multiple_alpha into 1 new EE task ucc_ee_task_reduce_strided
Updates all the TL/UCP collectives to use executor (when required) instead of MC (blocking reductions are still blocking so far, will be changed in separate PR)

## Why ?
We needed a way to add new varieties of EE tasks (another iface for reduction for TL/SHM, extended COPY tasks, etc)
Separation of reduce functionality from MC (memory component) to EC

samnordmann commented on src/components/ec/cuda/kernel/ec_cuda_reduce.cu:
```diff
@@ -0,0 +1,336 @@
+            break;                                                             \
+        default:                                                               \
+            for (i = start; i < count; i += step) {                            \
+                d[i] = _OP(s1[i], s2[i]);                                      \
```
Same remark as for line 50

samnordmann commented on src/components/ec/cuda/kernel/ec_cuda_reduce.cu:
```diff
@@ -0,0 +1,336 @@
+            break;                                                             \
+        case 3:                                                                \
+            for (i = start; i < count; i += step) {                            \
+                d[i] = _OP##_4(s1[i], s2[i], s2[i + ld], s2[i + 2 * ld]);      \
```
Would it be preferable to move the striding logic to ucc_dt_reduce_strided in ucc_dt_reduce.h (as proposed in a previous comment)? Maybe we would avoid duplicating CUDA_REDUCE_WITH_OP for the strided version

vspetrov replied:
not sure what do you mean by "move striding logic to ucc_dt_reduce_striedd" ? DO you mean converting to default reduce ? Then, again, the answer is that we want to keep task_reduce_strided for better signalling in ee_cuda.

samnordmann replied:
yes, this is what I meant. Ok for the explanation

vspetrov commented on src/components/ec/base/ucc_ec_base.h:
```diff
@@ -63,10 +63,9 @@ enum ucc_ee_executor_params_field {
-    UCC_EE_EXECUTOR_TASK_TYPE_REDUCE_MULTI       = UCC_BIT(2),
-    UCC_EE_EXECUTOR_TASK_TYPE_REDUCE_MULTI_ALPHA = UCC_BIT(3),
+    UCC_EE_EXECUTOR_TASK_REDUCE         = UCC_BIT(0),
+    UCC_EE_EXECUTOR_TASK_REDUCE_STRIDED = UCC_BIT(1),
```
the reason we keep it separate is that with strided we can describe task with multiple vectors using just 3 values. Keeping task_args as small is possible is critical for cpu->gpu signalling. So, it is optimization.

vspetrov commented on src/components/ec/cpu/ec_cpu.c:
```diff
@@ -112,49 +112,54 @@ ucc_status_t ucc_cpu_executor_task_post(ucc_ee_executor_t *executor,
+        ucc_eee_task_reduce_t tr;
+        int                   i;
+
+        if (n_srcs <= UCC_EE_EXECUTOR_NUM_BUFS) {
```
see comment above: we want to have strided for the more optimized signalling in cuda executor

vspetrov commented on src/components/ec/cpu/ec_cpu_reduce.c:
```diff
@@ -0,0 +1,323 @@
+            return UCC_ERR_NOT_SUPPORTED;                                      \
+        }                                                                      \
+        if (flags & UCC_EEE_TASK_FLAG_REDUCE_WITH_ALPHA) {                     \
+            VEC_OP(d, _count, task->alpha);                                    \
```
Yes, there would be less "loads". But as you mentioned the code will be much more complex while we still not see much request for UCC_OP_AVG (so it is rare corner case).

vspetrov commented on src/components/ec/cpu/ec_cpu_reduce.c:
```diff
@@ -0,0 +1,323 @@
+
+#define DO_DT_REDUCE_BFLOAT16(_srcs, _dst, _op, _count, _n_srcs)               \
+    do {                                                                       \
+        float _a = (flags & UCC_EEE_TASK_FLAG_REDUCE_WITH_ALPHA) ? task->alpha \
```
with bfloat16 on CPU we already pay super costly bfloat16<->float conversion on each step. so, i think this multiplication is negligible. I can do the same "if" in the end though, to be similar to other dtypes.

vspetrov commented:
@edgargabriel plz have a look. This PR requires the corresponding changes for MC/EC/ROCM. You mentioned that you can prep a PR against my branch so that everything goes to upstream/master together

edgargabriel commented:
@vspetrov I will work on it, I will try to have it ready by the end of the week

edgargabriel commented:
> @edgargabriel plz have a look. This PR requires the corresponding changes for MC/EC/ROCM. You mentioned that you can prep a PR against my branch so that everything goes to upstream/master together

@vspetrov I filed the pr against your branch, please have a look at let me know. (https://github.com/vspetrov/ucc/pull/3). Not entirely sure why it wants to merger your commit as well as the new stuff as part of the pr (I did use your branch as the starting point of my branch), but we'll find a solution to resolve this, the code is fundamentally there and tested.

vspetrov commented on src/components/ec/base/ucc_ec_base.h:
```diff
@@ -77,6 +76,44 @@ typedef struct ucc_ee_executor_params {
+
+typedef struct ucc_eee_task_reduce_strided {
+    ucc_datatype_t     dt;
+    ucc_reduction_op_t op;
```
i've reorganized a bit reducing from 72 bytes to 64. Still the size of ucc_eee_task_args will be defined by the ucc_ee_task_copy_multi (under union) which is 152 currently. 

Still if we optimize signalling (pidx is int, if we split it into 16bits task type and 16 bits actual pidx) then we can read only sizeof(task_reduce_strided) inside executor kernel

vspetrov commented:
@edgargabriel could you plz rebase your PR on top of latest changes in the branch. I've addressed comments and now it is incompatible with you changes. I assume you have changes in a separate PR so should be easy to rebase

edgargabriel commented:
> @edgargabriel could you plz rebase your PR on top of latest changes in the branch. I've addressed comments and now it is incompatible with you changes. I assume you have changes in a separate PR so should be easy to rebase

@vspetrov done

vspetrov commented on src/components/ec/cuda/kernel/ec_cuda_reduce.cu:
```diff
@@ -0,0 +1,336 @@
+#if CUDART_VERSION >= 11000
+    case UCC_DT_BFLOAT16:
+        ucc_assert(2 == sizeof(__nv_bfloat16));
+        DT_REDUCE_FLOAT(__nv_bfloat16, task, op, stream, bk, th);
```
I checked. We don't need. nvcc will generate bf16<->f32 conversion for arch < 80.

Sergei-Lebedev commented on src/components/ec/cuda/kernel/ec_cuda_reduce.cu:
```diff
@@ -0,0 +1,340 @@
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+    CUDA_CHECK(cudaGetLastError());
+    CUDA_CHECK(cudaStreamSynchronize(stream));
```
cudaStreamSynchronize is not needed, if executor is the only interface for reductions then ucc_executor_task_test should indicate about completion. Removing stream synchronize will make reduce async

vspetrov replied:
done

manjugv commented on src/components/ec/base/ucc_ec_base.h:
```diff
@@ -77,32 +76,83 @@ typedef struct ucc_ee_executor_params {
+   then the sources are taken from "srcs" and n_srcs can not exceed
+   UCC_EE_EXECUTOR_NUM_BUFS.
+
+   If UCC_EEE_TASK_FLAG_REDUCE_SRCS_EXT flag IS set on task_args
```
IS -> is

manjugv commented on src/components/ec/base/ucc_ec_base.h:
```diff
@@ -77,32 +76,83 @@ typedef struct ucc_ee_executor_params {
+        void **srcs_ext;
+    };
+    size_t   count;
+    double   alpha;
```
align
